core/enumerator/product_spec.rb

```diff
@@ -14,9 +37,33 @@
       elems.should == [[1, "X"], [1, "Y"], [2, "X"], [2, "Y"]]
     end
 
+    it "accepts infinite enumerator and a finite enumerable" do
+      enum = Enumerator.product(1.., "A".."C")
+      enum.size.should == Float::INFINITY
+      enum.take(4).should == [[1, "A"], [1, "B"], [1, "C"], [2, "A"]]
+    end
+
     it "accepts infinite enumerator and uknown enumerator" do
       enum = Enumerator.product(1.., Enumerator.new { |y| y << "X" << "Y" })
+      enum.size.should == Float::INFINITY
       enum.take(4).should == [[1, "X"], [1, "Y"], [2, "X"], [2, "Y"]]
     end
+
+    it "accepts finite enumerator and unknown anumerator" do
+      enum = Enumerator.product(1..2, Enumerator.new { |y| y << "X" << "Y" })
+      enum.size.should == nil
+      enum.to_a.should == [[1, "X"], [1, "Y"], [2, "X"], [2, "Y"]]
+    end
+
+    it "reject keyword arguments" do
+      -> {
+        Enumerator.product(1..3, foo: 1, bar: 2)
+      }.should raise_error(ArgumentError)
+    end
+
+    it "can be iterated with #each_entry" do
+      enum = Enumerator.product(1..2, ["A", "B"])
+      enum.each_entry.to_a.should == [[1, "A"], [1, "B"], [2, "A"], [2, "B"]]
+    end
```

Comment:
**minor:** Actually it's testing interface of the `Product` class. Moreover `Product` class inherits `#each_entry` from the `Enumerator` class and this method is defined in `Enumerable`... So there is nothing to check actually.

As far as I understand `#each_entry` method is called on arguments. So it's probably that it is the only requirement for the arguments. And they can be instances of any class, not only enumerable/enumerator, but any class that just implements the `#each_entry` method.

It's common to check such behaviour with `Object.new` e.g. we can check it in the following way:

```ruby
object = Object.new
def object.each_entry
  yield 1
  yield 2
end

Enumerator.product(object, ["A", "B"]).to_a.should == [[1, "A"], [1, "B"], [2, "A"], [2, "B"]]
```